### DISCUSSION TITLE
Multiple machine available for multiple job to work parallel: job shop problem

### DISCUSSION THREAD

**Discussion by Nutchareessw** [google/or-tools#2868] [CP-SAT questions] [answered]
**What language and solver does this apply to?**
Python / CP-SAT

**Describe the problem you are trying to solve.**
I am newbie using or-tools. Base on the job shop problem and dataset is
jobs_data = [  # task = (machine_id, processing_time).
    [(0, 2), (1, 2), (2, 2)],  # Job0
    [(0, 2), (2, 1), (1, 4)],  # Job1
    [(1, 2), (2, 3)]  # Job2
]

but in my case if  I will have 3 machines type 0 available [ for example machine _id for type 0 is 01, 02, 03], 2 machines type 1 (11, 12 )and 1 machine type 2 (21) all machine can run in the same time. 
For example in the dataset job0 and job1 request to work on machine 0. both jobs can run on machine 01 and 02 as well as job2 can run on machine 1 in the same time. So I will finish 3 jobs (0,2) (0,2) (1,2) in the first 2 day, then the next job order can start work. I only focus on NOoverlapping constraint. 

I have gone through the job shop example, but still don't understand much how to model the above issue using the CP solver? 
Any help is appreciated.

**Comment by lperron** [comment, marked as answer]
https://github.com/google/or-tools/blob/stable/examples/python/flexible_job_shop_sat.py

**Reply by Nutchareessw** [reply]
thanks for quick reply! I have gone through the link and as I'm very new with or-tools, I still don't get how to create a variable to tell the model that
I have 3 machines of type 0 available, 2 machines of type 1 available and 1 machine type 2 available and all machines are available to run but one machine can do only one task in a time. Can you pls give me any hint?

my outcome should similar to this pic.
![Capture](https://user-images.githubusercontent.com/79275381/138661248-4659daa9-156d-4acb-b27b-64fc1a9824a9.PNG)

**Reply by Mizux** [reply]
do you mean machines `0,0` `1,0` and `2,0` are actually the same machine (i.e. a machine we can performs the 3 tasks type but only one at a time?) ?
note: let's call it a "multi-tasks" machine

**Reply by Nutchareessw** [reply]
@Mizux
Sorry for my unclear explanation. 
this case I assume that I have 3 types of machine and one type of machine can have 1 or more than 1 machine.  For example, I have 3 machine of machine type 0. 
(machine_type, machine_id)
(0,0) mean machine type 0 machine_id 0
(0,1) mean machine type 0 machine_id 1
(0,2) mean machine type 0 machine_id 2
(1,0) mean machine type 1 machine_id 0
(1,1) mean machine type 1 machine_id 1
(2,0) mean machine type 2 machine_id 0

-------------------------------------------------

(task = (machine_type, processing_time  ))
[(0, 2), (1, 2), (2, 2)] #job0 
[(0, 2), (2, 1), (1, 4)] #job1
For example, job 0 need to perform 3 tasks, also job1 need to perform 3 tasks.
job 0 task 0 requires to run on machine type 0 for 2 days, and job 1 task 0 requires to run on machine type 0 for 2 days.
As, I have 3 machines of machine type 0 available, So job0 task 0 and job1 task 0 can run on any machine type 0 (same as in the picture above), but 1 one machine can perform only one task in a time. 

How to define in the model that I have mutiple type of machine and one type can have mutiple machines and jobs are not overlapping.

**Reply by Mizux** [reply]
So If I follow you correctly you have 6 machines, three of type 0, two of type 2 and one of type 3. so you it should be better to use:
```py
# (machine_type, machine_id)
(0,0) # mean machine type 0 machine_id 0
(0,1) # mean machine type 0 machine_id 1
(0,2) # mean machine type 0 machine_id 2
(1,3) # mean machine type 1 machine_id 3
(1,4) # mean machine type 1 machine_id 4
(2,5) # mean machine type 2 machine_id 5
```

After as laurent point you you have the **flexible** jobshop example which show you how to have alternative tasks
so in your example and using my notation I would have:
```python
# Data part.
    jobs = [  # task = (processing_time, machine_id)
        [  # Job 0
            [(2, 0), (2, 1), (2, 2)],  # task 0 (machine type 0) with 3 alternatives
            [(1, 3), (1, 4)],  # task 1 (machine type 1) with 2 alternatives
            [(2, 5)],  # task 2 (machine type 2) with 1 alternatives
        ],
        [  # Job 1
            [(2, 0), (2, 1), (2, 2)],# task 0 (machine type 0) with 3 alternatives
            [(1, 5)],# task 1 (machine type 2) with 1 alternatives
            [(4, 3), (4, 4)],# task 3 (machine type 1) with 2 alternatives
        ],
    ]
```

**Reply by Nutchareessw** [reply]
Thanks Mizux, this solution helps me a lot. 
Is it possible to solve the same problem if I give separate dataset between job_data and machine_data like below in the model?? 
if yes, it can solve, is any hint you can give me? 
jobs_data = [  # task = (machine_type, processing_time).
    [(0, 3), (1, 2), (2, 2)],  # Job0
    [(0, 2), (2, 1), (1, 4)],  # Job1
    [(1, 4), (2, 3)]  # Job2
]

#(machine_type, capacity)   total 6 machines, 3 of type 0, 2 of type 2 and one of type 3
machine = [[0,3], [1,2], [2,1]]

**Reply by Mizux** [reply]
basically you can do whatever you want provide you are fluent enough in python, i.e. you just need to adapt the whole sample to your need, the sat solver don't impose you any structures for your model.
TLDR you just need to map your model to a bunch of IntVar/BoolVar and IntervalVar + few constraints (no overlap, etc...)
